numpyro/util.py

```diff
@@ -104,3 +109,73 @@ def scan(f, a, bs):
         return tree_unflatten(tree_def, as_)
     else:
         return lax.scan(f, a, bs)
+
+
+def tscan(f, a, bs, fields=(0,)):
```

Comment:
Maybe it's worth having a discussion with the JAX devs to see if they are open to integrating this upstream?

Reply:
The problem is this `fields` is used for the order in unflatten tree. So if `a` contains dict of dicts, users have to figure out which indices they have to put into the fields by looking at treedef. This can be generalized but I am not sure how to do it properly.

Reply:
See my comment below on taking a transformation function instead of `fields` that would work more generally. e.g. in our case, the transformation function might simply be `lambda state: state.z`.

Reply:
I'm fine investigating this later too, so this doesn't have to be part of this PR!